### PR TITLE
ADHOC: Volume fsck use a time cutoff param

### PR DESCRIPTION
**Pull request by lyang24**
# What problem are we solving?
https://github.com/seaweedfs/seaweedfs/issues/2578 
when we are copying files and run volumn.fsck at the same time it could lead to results where file chucks appears on volume  server but filer do not contains the file info.

# How are we solving the problem?
we added a cutoff time requirement default to 30 mins on volume.fsck command. we will ignore files created in the last 30 mins in idx files. we ran binary search to query needle meta data to find the last vaild needle that that needle's last modified time <= cutoff time requirement.

# How is the PR tested?

# Checks
- [ ] I have added unit tests if possible.
- [ ] I will add related wiki document changes and link to this PR after merging.

### CODE REVIEW

**Comment by kmlebedev** [comment]
@lyang24 Can it be possible to fix this in one? https://github.com/seaweedfs/seaweedfs/issues/3582

**Comment by chrislusf** [comment]
> @lyang24 Can it be possible to fix this in one? #3582

Unrelated.

File: weed/storage/idx/binary_search.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package idx
+
+import (
+	"github.com/seaweedfs/seaweedfs/weed/storage/types"
+)
+
+// firstInvalidIndex find the first index that do not satisfy the validation function's requirement.
+func FirstInvalidIndex(bytes []byte, validation func(key types.NeedleId, offset types.Offset, size types.Size) (bool, error)) (int, error) {
```
Comment (chrislusf) [inline weed/storage/idx/binary_search.go]: can you rename "validation" to "lessThanFn" or "greaterThanFn"?

Comment (lyang24) [reply weed/storage/idx/binary_search.go]: fixed

File: weed/shell/command_volume_fsck.go [hunk 2]
```diff
@@ -72,6 +75,7 @@ func (c *commandVolumeFsck) Do(args []string, commandEnv *CommandEnv, writer io.
 	c.forcePurging = fsckCommand.Bool("forcePurging", false, "delete missing data from volumes in one replica used together with applyPurging")
 	purgeAbsent := fsckCommand.Bool("reallyDeleteFilerEntries", false, "<expert only!> delete missing file entries from filer if the corresponding volume is missing for any reason, please ensure all still existing/expected volumes are connected! used together with findMissingChunksInFiler")
 	tempPath := fsckCommand.String("tempPath", path.Join(os.TempDir()), "path for temporary idx files")
+	cutoffTimeAgo := fsckCommand.Uint64("cutoffTimeAgo", uint64(time.Now().Add(-time.Minute*30).Unix()), "the offset of filtering files on volume server")
```
Comment (chrislusf) [inline weed/shell/command_volume_fsck.go]: use `time.Duration` similar to this https://github.com/seaweedfs/seaweedfs/blob/358ccb5e0eb2fd0d5f7f1bd6259dc5950df76526/weed/command/filer_backup.go#L35

Comment (lyang24) [reply weed/shell/command_volume_fsck.go]: fixed

File: weed/shell/command_volume_fsck.go [hunk 3]
```diff
@@ -75,7 +75,7 @@ func (c *commandVolumeFsck) Do(args []string, commandEnv *CommandEnv, writer io.
 	c.forcePurging = fsckCommand.Bool("forcePurging", false, "delete missing data from volumes in one replica used together with applyPurging")
 	purgeAbsent := fsckCommand.Bool("reallyDeleteFilerEntries", false, "<expert only!> delete missing file entries from filer if the corresponding volume is missing for any reason, please ensure all still existing/expected volumes are connected! used together with findMissingChunksInFiler")
 	tempPath := fsckCommand.String("tempPath", path.Join(os.TempDir()), "path for temporary idx files")
-	cutoffTimeAgo := fsckCommand.Uint64("cutoffTimeAgo", uint64(time.Now().Add(-time.Minute*30).Unix()), "the offset of filtering files on volume server")
+	cutoffTimeAgo := fsckCommand.Duration("cutoffTimeAgo", 30*time.Minute, "the offset of filtering files on volume server")
```
Comment (chrislusf) [inline weed/shell/command_volume_fsck.go]:
```suggestion
	cutoffTimeAgo := fsckCommand.Duration("cutoffTimeAgo", 5*time.Minute, "only include entries  on volume servers before this cutoff time to check orphan chunks")
```

**Pull request merged**